File: syndrome/core/partition.py
"""
Partition Geometry Module

Implements partition coordinates (n, ℓ, m, s) and related operations
derived from bounded phase space and categorical observation axioms.

The partition capacity C(n) = 2n² emerges as geometric necessity from
spherical symmetry of bounded phase space.
"""

from dataclasses import dataclass
from typing import Tuple, List
import numpy as np


@dataclass(frozen=True)
class PartitionCoord:
    """
    Partition coordinates in categorical space.

    Attributes:
        n: Depth coordinate (n >= 1)
        ell: Complexity coordinate (0 <= ell <= n-1)
        m: Orientation coordinate (-ell <= m <= ell)
        s: Chirality coordinate (s in {-0.5, +0.5})
    """
    n: int
    ell: int
    m: int
    s: float

    def __post_init__(self) -> None:
        """Validate partition coordinates."""
        if not is_valid_partition_coord(self.n, self.ell, self.m, self.s):
            raise ValueError(
                f"Invalid partition coordinates: n={self.n}, ℓ={self.ell}, "
                f"m={self.m}, s={self.s}"
            )

    def to_tuple(self) -> Tuple[int, int, int, float]:
        """Convert to tuple representation."""
        return (self.n, self.ell, self.m, self.s)

    def to_array(self) -> np.ndarray:
        """Convert to numpy array for numerical operations."""
        return np.array([self.n, self.ell, self.m, self.s], dtype=np.float64)


def is_valid_partition_coord(n: int, ell: int, m: int, s: float) -> bool:
    """
    Check if partition coordinates are valid.

    Valid coordinates satisfy:
    - n >= 1 (at least one partition)
    - 0 <= ell <= n-1 (complexity bounded by depth)
    - -ell <= m <= ell (orientation bounded by complexity)
    - s in {-0.5, +0.5} (chirality is binary)

    Args:
        n: Depth coordinate
        ell: Complexity coordinate
        m: Orientation coordinate
        s: Chirality coordinate

    Returns:
        True if coordinates are valid, False otherwise
    """
    if n < 1:
        return False
    if ell < 0 or ell > n - 1:
        return False
    if m < -ell or m > ell:
        return False
    if s not in (-0.5, 0.5):
        return False
    return True


def partition_capacity(n: int) -> int:
    """
    Compute partition capacity at depth n.

    The capacity C(n) = 2n² follows from:
    - Sum over ell from 0 to n-1 of (2*ell + 1) orientations
    - Factor of 2 for chirality

    This yields the sequence: 2, 8, 18, 32, 50, 72, 98, ...
    which matches electron shell capacities in atomic physics.

    Args:
        n: Partition depth (n >= 1)

    Returns:
        Capacity C(n) = 2n²

    Raises:
        ValueError: If n < 1
    """
    if n < 1:
        raise ValueError(f"Partition depth must be >= 1, got {n}")
    return 2 * n * n


def categorical_distance(
    sigma1: PartitionCoord,
    sigma2: PartitionCoord
) -> float:
    """
    Compute categorical distance between partition states.

    The categorical distance is defined as Euclidean distance in
    partition coordinate space:

    d_cat = sqrt((n1-n2)² + (ℓ1-ℓ2)² + (m1-m2)² + (s1-s2)²)

    IMPORTANT: Categorical distance is independent of:
    - Spatial distance (d_cat ⊥ d_spatial)
    - Optical opacity (∂d_cat/∂τ = 0)

    This enables opacity-independent measurement of states behind
    arbitrary optical barriers.

    Args:
        sigma1: First partition state
        sigma2: Second partition state

    Returns:
        Categorical distance d_cat >= 0
    """
    arr1 = sigma1.to_array()
    arr2 = sigma2.to_array()
    return float(np.linalg.norm(arr1 - arr2))


def categorical_distance_raw(
    n1: int, ell1: int, m1: int, s1: float,
    n2: int, ell2: int, m2: int, s2: float
) -> float:
    """
    Compute categorical distance from raw coordinates.

    Args:
        n1, ell1, m1, s1: First partition state coordinates
        n2, ell2, m2, s2: Second partition state coordinates

    Returns:
        Categorical distance d_cat >= 0
    """
    return np.sqrt(
        (n1 - n2)**2 +
        (ell1 - ell2)**2 +
        (m1 - m2)**2 +
        (s1 - s2)**2
    )


def enumerate_partition_states(n: int) -> List[PartitionCoord]:
    """
    Enumerate all partition states at depth n.

    Args:
        n: Partition depth

    Returns:
        List of all C(n) = 2n² partition states at depth n
    """
    states = []
    for ell in range(n):
        for m in range(-ell, ell + 1):
            for s in (-0.5, 0.5):
                states.append(PartitionCoord(n=n, ell=ell, m=m, s=s))
    return states


def address_to_value(
    address: List[int],
    range_min: float = 0.0,
    range_max: float = 1.0
) -> float:
    """
    Resolve ternary categorical address to physical value.

    An address [t_0, t_1, ..., t_{k-1}] with t_i in {0, 1, 2}
    resolves to:

    value = range_min + Σ_i (t_i / 3^{i+1}) * (range_max - range_min)

    Args:
        address: List of ternary digits
        range_min: Minimum of physical range
        range_max: Maximum of physical range

    Returns:
        Resolved physical value
    """
    if not address:
        return (range_min + range_max) / 2

    delta = range_max - range_min
    value = 0.0
    for i, t in enumerate(address):
        if t not in (0, 1, 2):
            raise ValueError(f"Address digit must be in {{0, 1, 2}}, got {t}")
        value += t / (3 ** (i + 1))

    return range_min + value * delta


def value_to_address(
    value: float,
    depth: int,
    range_min: float = 0.0,
    range_max: float = 1.0
) -> List[int]:
    """
    Convert physical value to ternary categorical address.

    Args:
        value: Physical value to encode
        depth: Depth of address (number of ternary digits)
        range_min: Minimum of physical range
        range_max: Maximum of physical range

    Returns:
        Ternary address as list of digits
    """
    if value < range_min or value > range_max:
        raise ValueError(
            f"Value {value} outside range [{range_min}, {range_max}]"
        )

    # Normalize to [0, 1]
    normalized = (value - range_min) / (range_max - range_min)

    address = []
    remaining = normalized
    for _ in range(depth):
        digit = int(remaining * 3)
        digit = min(digit, 2)  # Clamp to valid range
        address.append(digit)
        remaining = remaining * 3 - digit

    return address

File: syndrome-rs/src/partition.rs
//! Partition Geometry Module
//!
//! Implements partition coordinates `(n, ℓ, m, s)` and related operations
//! derived from bounded phase space and categorical observation axioms.
//!
//! The partition capacity `C(n) = 2n²` emerges as geometric necessity from
//! spherical symmetry of bounded phase space.

use serde::{Deserialize, Serialize};
use crate::{Result, SyndromeError};

/// Chirality coordinate (spin-like binary value).
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize)]
pub enum Chirality {
    /// Positive chirality (+1/2)
    Plus,
    /// Negative chirality (-1/2)
    Minus,
}

impl Chirality {
    /// Get the numerical value of chirality.
    #[inline]
    pub fn value(self) -> f64 {
        match self {
            Chirality::Plus => 0.5,
            Chirality::Minus => -0.5,
        }
    }

    /// Create chirality from numerical value.
    pub fn from_value(v: f64) -> Option<Self> {
        if (v - 0.5).abs() < 1e-10 {
            Some(Chirality::Plus)
        } else if (v + 0.5).abs() < 1e-10 {
            Some(Chirality::Minus)
        } else {
            None
        }
    }
}

/// Partition coordinates in categorical space.
///
/// # Coordinates
/// - `n`: Depth coordinate (n ≥ 1)
/// - `ell`: Complexity coordinate (0 ≤ ℓ ≤ n-1)
/// - `m`: Orientation coordinate (-ℓ ≤ m ≤ ℓ)
/// - `s`: Chirality coordinate (s ∈ {-0.5, +0.5})
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize)]
pub struct PartitionCoord {
    n: i32,
    ell: i32,
    m: i32,
    s: Chirality,
}

impl PartitionCoord {
    /// Create a new partition coordinate with validation.
    ///
    /// # Arguments
    /// * `n` - Depth coordinate (must be ≥ 1)
    /// * `ell` - Complexity coordinate (must be 0 ≤ ℓ ≤ n-1)
    /// * `m` - Orientation coordinate (must be -ℓ ≤ m ≤ ℓ)
    /// * `s` - Chirality coordinate
    ///
    /// # Errors
    /// Returns `SyndromeError::InvalidPartitionCoord` if constraints are violated.
    pub fn new(n: i32, ell: i32, m: i32, s: Chirality) -> Result<Self> {
        if !is_valid_partition_coord(n, ell, m, s.value()) {
            return Err(SyndromeError::InvalidPartitionCoord {
                n,
                ell,
                m,
                s: s.value(),
            });
        }
        Ok(Self { n, ell, m, s })
    }

    /// Create without validation (for internal use).
    #[inline]
    pub(crate) fn new_unchecked(n: i32, ell: i32, m: i32, s: Chirality) -> Self {
        Self { n, ell, m, s }
    }

    /// Get the depth coordinate.
    #[inline]
    pub fn n(&self) -> i32 {
        self.n
    }

    /// Get the complexity coordinate.
    #[inline]
    pub fn ell(&self) -> i32 {
        self.ell
    }

    /// Get the orientation coordinate.
    #[inline]
    pub fn m(&self) -> i32 {
        self.m
    }

    /// Get the chirality coordinate.
    #[inline]
    pub fn chirality(&self) -> Chirality {
        self.s
    }

    /// Get chirality as numerical value.
    #[inline]
    pub fn s(&self) -> f64 {
        self.s.value()
    }

    /// Convert to array representation for numerical operations.
    #[inline]
    pub fn to_array(&self) -> [f64; 4] {
        [self.n as f64, self.ell as f64, self.m as f64, self.s.value()]
    }

    /// Get the partition capacity at this depth.
    ///
    /// Returns `C(n) = 2n²`.
    #[inline]
    pub fn capacity(&self) -> i32 {
        partition_capacity(self.n)
    }
}

/// Check if partition coordinates are valid.
///
/// Valid coordinates satisfy:
/// - n ≥ 1 (at least one partition)
/// - 0 ≤ ℓ ≤ n-1 (complexity bounded by depth)
/// - -ℓ ≤ m ≤ ℓ (orientation bounded by complexity)
/// - s ∈ {-0.5, +0.5} (chirality is binary)
#[inline]
pub fn is_valid_partition_coord(n: i32, ell: i32, m: i32, s: f64) -> bool {
    if n < 1 {
        return false;
    }
    if ell < 0 || ell > n - 1 {
        return false;
    }
    if m < -ell || m > ell {
        return false;
    }
    if (s - 0.5).abs() > 1e-10 && (s + 0.5).abs() > 1e-10 {
        return false;
    }
    true
}

/// Compute partition capacity at depth n.
///
/// The capacity `C(n) = 2n²` follows from:
/// - Sum over ℓ from 0 to n-1 of (2ℓ + 1) orientations
/// - Factor of 2 for chirality
///
/// This yields the sequence: 2, 8, 18, 32, 50, 72, 98, ...
/// which matches electron shell capacities in atomic physics.
///
/// # Panics
/// Panics if `n < 1`.
#[inline]
pub fn partition_capacity(n: i32) -> i32 {
    assert!(n >= 1, "Partition depth must be >= 1, got {}", n);
    2 * n * n
}

/// Compute categorical distance between partition states.
///
/// The categorical distance is defined as Euclidean distance in
/// partition coordinate space:
///
/// `d_cat = sqrt((n₁-n₂)² + (ℓ₁-ℓ₂)² + (m₁-m₂)² + (s₁-s₂)²)`
///
/// **IMPORTANT**: Categorical distance is independent of:
/// - Spatial distance (`d_cat ⊥ d_spatial`)
/// - Optical opacity (`∂d_cat/∂τ = 0`)
///
/// This enables opacity-independent measurement of states behind
/// arbitrary optical barriers.
#[inline]
pub fn categorical_distance(sigma1: &PartitionCoord, sigma2: &PartitionCoord) -> f64 {
    categorical_distance_raw(
        sigma1.n, sigma1.ell, sigma1.m, sigma1.s(),
        sigma2.n, sigma2.ell, sigma2.m, sigma2.s(),
    )
}

/// Compute categorical distance from raw coordinates.
#[inline]
pub fn categorical_distance_raw(
    n1: i32, ell1: i32, m1: i32, s1: f64,
    n2: i32, ell2: i32, m2: i32, s2: f64,
) -> f64 {
    let dn = (n1 - n2) as f64;
    let dell = (ell1 - ell2) as f64;
    let dm = (m1 - m2) as f64;
    let ds = s1 - s2;
    (dn * dn + dell * dell + dm * dm + ds * ds).sqrt()
}

/// Enumerate all partition states at depth n.
///
/// Returns all `C(n) = 2n²` partition states at the given depth.
pub fn enumerate_partition_states(n: i32) -> Vec<PartitionCoord> {
    assert!(n >= 1, "Partition depth must be >= 1");
    let capacity = partition_capacity(n) as usize;
    let mut states = Vec::with_capacity(capacity);

    for ell in 0..n {
        for m in -ell..=ell {
            states.push(PartitionCoord::new_unchecked(n, ell, m, Chirality::Minus));
            states.push(PartitionCoord::new_unchecked(n, ell, m, Chirality::Plus));
        }
    }

    states
}

/// Resolve ternary categorical address to physical value.
///
/// An address `[t_0, t_1, ..., t_{k-1}]` with `t_i ∈ {0, 1, 2}` resolves to:
///
/// `value = range_min + Σ_i (t_i / 3^{i+1}) * (range_max - range_min)`
///
/// # Errors
/// Returns `SyndromeError::InvalidAddressDigit` if any digit is not in {0, 1, 2}.
pub fn address_to_value(address: &[i32], range_min: f64, range_max: f64) -> Result<f64> {
    if address.is_empty() {
        return Ok((range_min + range_max) / 2.0);
    }

    let delta = range_max - range_min;
    let mut value = 0.0;

    for (i, &t) in address.iter().enumerate() {
        if t < 0 || t > 2 {
            return Err(SyndromeError::InvalidAddressDigit { digit: t });
        }
        value += (t as f64) / 3.0_f64.powi((i + 1) as i32);
    }

    Ok(range_min + value * delta)
}

/// Convert physical value to ternary categorical address.
///
/// # Errors
/// Returns `SyndromeError::OutOfBounds` if value is outside range.
pub fn value_to_address(
    value: f64,
    depth: usize,
    range_min: f64,
    range_max: f64,
) -> Result<Vec<i32>> {
    if value < range_min || value > range_max {
        return Err(SyndromeError::OutOfBounds {
            value,
            min: range_min,
            max: range_max,
        });
    }

    // Normalize to [0, 1]
    let normalized = (value - range_min) / (range_max - range_min);

    let mut address = Vec::with_capacity(depth);
    let mut remaining = normalized;

    for _ in 0..depth {
        let digit = (remaining * 3.0).floor() as i32;
        let digit = digit.min(2); // Clamp to valid range
        address.push(digit);
        remaining = remaining * 3.0 - digit as f64;
    }

    Ok(address)
}

/// Compute precision of address at given depth.
///
/// Precision = 1 / 3^depth
#[inline]
pub fn address_precision(depth: usize) -> f64 {
    1.0 / 3.0_f64.powi(depth as i32)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_partition_capacity() {
        assert_eq!(partition_capacity(1), 2);
        assert_eq!(partition_capacity(2), 8);
        assert_eq!(partition_capacity(3), 18);
        assert_eq!(partition_capacity(4), 32);
        assert_eq!(partition_capacity(5), 50);
    }

    #[test]
    fn test_valid_coords() {
        assert!(PartitionCoord::new(1, 0, 0, Chirality::Plus).is_ok());
        assert!(PartitionCoord::new(2, 1, -1, Chirality::Minus).is_ok());
        assert!(PartitionCoord::new(3, 2, 2, Chirality::Plus).is_ok());
    }

    #[test]
    fn test_invalid_coords() {
        // n < 1
        assert!(PartitionCoord::new(0, 0, 0, Chirality::Plus).is_err());
        // ell >= n
        assert!(PartitionCoord::new(2, 2, 0, Chirality::Plus).is_err());
        // |m| > ell
        assert!(PartitionCoord::new(2, 1, 2, Chirality::Plus).is_err());
    }

    #[test]
    fn test_enumerate_states() {
        let states = enumerate_partition_states(2);
        assert_eq!(states.len(), 8);
    }

    #[test]
    fn test_categorical_distance_self() {
        let coord = PartitionCoord::new(2, 1, 0, Chirality::Plus).unwrap();
        assert!((categorical_distance(&coord, &coord)).abs() < 1e-10);
    }

    #[test]
    fn test_categorical_distance_chirality() {
        let c1 = PartitionCoord::new(1, 0, 0, Chirality::Plus).unwrap();
        let c2 = PartitionCoord::new(1, 0, 0, Chirality::Minus).unwrap();
        assert!((categorical_distance(&c1, &c2) - 1.0).abs() < 1e-10);
    }

    #[test]
    fn test_address_resolution() {
        let address = vec![1, 1, 1];
        let value = address_to_value(&address, 0.0, 1.0).unwrap();
        assert!((value - 0.481481).abs() < 0.001);
    }

    #[test]
    fn test_value_to_address() {
        let value = 0.5;
        let address = value_to_address(value, 5, 0.0, 1.0).unwrap();
        let resolved = address_to_value(&address, 0.0, 1.0).unwrap();
        assert!((value - resolved).abs() < address_precision(5));
    }

    #[test]
    fn test_address_precision() {
        assert!((address_precision(1) - 1.0/3.0).abs() < 1e-10);
        assert!((address_precision(2) - 1.0/9.0).abs() < 1e-10);
        assert!((address_precision(3) - 1.0/27.0).abs() < 1e-10);
    }
}
